md/configure.md
## I STRONGLY ADVISE YOU TO WAIT FOR THE RELEASE OF THE PROGRAM AS ALL THIS INFORMATION MAY LOSE RELEVANCE AT ANY MOMENT
## It is already lost relevance xd. Wait for an update
## Configure
## config.json file
### parameters:
|parameter|type|description|possible values|default value|
|---|---|---|---|---|
|displays|List|list of displays|
|temp_path|String|path to templates.json|path|~/.config/rpaper/templates.json
|vars_path|String|path to color_variables.json|path|~/.config/rpaper/color_variables.json
|cache_dir|String|path to splited wallpapers|path|~/.cache/rpaper/Wallpapers
|scheme_file|String|path to color scheme file|path|~/.cache/rpaper/rwal/colors
|wall_command|String|command to set wallpapers to displays. supports formatting where {display} is a name of display and {image} is a path to splitted wallpaper|swaybg/swww command|swaybg -o {display} -i {image}
|resize_algorithm|String|image resize algorithm|Nearest CatmullRom Gaussian Lanczos3 Triangle|Triangle|
|cache_scheme|Boolean|responsible for creating color pallete|true/false|true
|set_templates|Boolean|responsible for changing configs|true/false|true
|cache_walls|Boolean|responsible for splitting wallpaper|true/false|true
|set_walls|Boolean|responsible for setting wallpaper|true/false|true
|imgp_change_contrast|Boolean|responsible for applying contrast to wallpaper|true/false|false
|imgp_change_brightness|Boolean|responsible for applying brightness to wallpaper|true/false|false
|imgp_change_huerotate|Boolean|responsible for applying huerotate to wallpaper|true/false|false
|imgp_change_blur|Boolean|responsible for applying blur to wallpaper|true/false|false
|imgp_contrast|Float|value to apply|-255 to 255|0
|imgp_brightness|Float|value to apply|-255 to 255|0
|imgp_huerotate|Number|value to apply|-255 to 255|0
|imgp_blur|Number|value to apply|0 to 255|0
|imgp_invert|Boolean|inverting wallpaper colors|true/false|false
|imgp_flip_h|Boolean|horizontaly flip wallpaper|true/false|false
|imgp_flip_v|Boolean|verticaly flip wallpaper|true/false|false
|rwal_cache_dir|String|path to cached color palettes|path|~/.cache/rpaper/rwal
|rwal_thumb_w|Number|width of thumbed wallpaper|2-any|100
|rwal_thumb_h|Number|height of thumbed wallpaper|2-any|100
|rwal_accent_color|Number|color from pallete from which bg and fg colors are generated|0-5|2
|rwal_clamp_min|Float|minimal brightness of colors in pallete|0-255|100
|rwal_clamp_max|Float|maximum brightness of colors in pallete|0-255|100
### displays:
|parameter|type|description|possible values|default value|
|---|---|---|---|---|
|name|String|name of display|any|none|
|width|Number|width of display|any|none|
|height|Number|height of display|any|none|
|margin-left|Number|x position of display|any|none|
|margin-top|Number|y position of display|any|none|
### example:
```json
{
  "displays": [
    {
      "name": "HDMI-A-1",
      "width": 1920,
      "height": 1080,
      "margin-left": 1080,
      "margin-top": 305
    },
    {
      "name": "DP-1",
      "width": 1080,
      "height": 1920,
      "margin-left": 0,
      "margin-top": 0
    }
  ],
  "temp_path": "~/.config/rpaper/templates.json",
  "vars_path": "~/.config/rpaper/color_variables.json",
  "cache_dir": "~/.cache/rpaper/Wallpapers",
  "scheme_file": "~/.cache/rpaper/rwal/colors",
  "wall_command": "swaybg -o {display} -i {image}",
  "resize_algorithm": "Lanczos3",

  "cache_scheme": true,
  "cache_walls": true,
  "set_templates": true,
  "set_walls": true,

  "imgp_change_contrast": false,
  "imgp_contrast": -15,
  "imgp_change_brightness": false,
  "imgp_brightness": -60,
  "imgp_change_huerotate": false,
  "imgp_huerotate": 200,
  "imgp_change_blur": false,
  "imgp_blur": 7.5,
  "imgp_invert": false,
  "imgp_flip_h": false,
  "imgp_flip_v": false,

  "rwal_cache_dir": "~/.cache/rpaper/rwal",
  "rwal_thumb_w": 210,
  "rwal_thumb_h": 210,
  "rwal_accent_color": 4,
  "rwal_clamp_min": 140.0,
  "rwal_clamp_max": 170.0
}
```
## templates.json
### parameters
|parameter|type|description|
|---|---|---|
|template_path|String|path to template config file|
|config_path|String|path to config config file|
|use_quotes|Boolean|param that wraps color in "": #001122 to "#001122"|
|use_sharps|Boolean|param that uses sharp before color: 001122 to #001122|
|opacity|String|param that is added after color: 001122 to 00112233|
|command|String|param that is execute command before&after changing config|
```
to use before-command  use || in command: killall waybar||waybar&
```
### example:
```json
[
  {
    "template_path": "~/.my_script_files/wallpaperpy_templates/alacritty.toml",
    "config_path": "~/.config/alacritty/alacritty.toml",
    "use_quotes": true,
    "use_sharps": true,
    "opacity": "",
    "command": ""
  },
  {
    "template_path": "~/.my_script_files/wallpaperpy_templates/waybar.css",
    "config_path": "~/.config/waybar/style.css",
    "use_quotes": false,
    "use_sharps": true,
    "opacity": "",
    "command": "killall waybar"
  }
]
```
## color_variables.json
### parameters
|parameter|type|description|
|---|---|---|
|name|String|name of color variable|
|value|Number|value of color from palette|
|brightness|Number|value which would be added to r,g,b of color|
|inverted|Boolean|invert color|
to create multiple brightness colors use {br} in color name. This would generate 20 more colors.
- Example:
```
"name": "(bg{br})" = (bgd1), (bgd2), (bgd3), (bgd4), (bgl1), (bgl2), (bgl3), (bgl4) etc.
```
d/l means darker/lighter, number is a number of tens added to the brightness of the original color.

### example:
```json
[
  {
    "name": "(bg{br})",
    "value": 15,
    "brightness": 10,
    "inverted": false
  },
  {
    "name": "(text)",
    "value": 0,
    "brightness": 0,
    "inverted": false
  },
  {
    "name": "(primary{br})",
    "value": 6,
    "brightness": 0
  }
]
```


daemon/src/daemon/config.rs
use std::fs::File;
use std::io::Read;

use serde_json::Value;

use crate::colorscheme::rwal::OrderBy;
use crate::expand_user;
use crate::wallpaper::display::Display;
use crate::wallpaper::display::ImageOperations;
use crate::colorscheme::template::Template;
use crate::colorscheme::rwal::RwalParams;
use crate::wallpaper::display::WCacheInfo;


#[derive(Clone)]
pub struct Config {
    pub displays: Option<Vec<Display>>,
    pub templates: Option<Vec<Template>>,
    pub wallpaper_set_command: Option<String>,
    pub resize_algorithm: Option<String>,
    pub rwal_params: Option<RwalParams>,
    pub image_operations: Option<ImageOperations>,
}

impl Config {
    pub fn new() -> Self {
        Config {
            displays: None,
            templates: None,
            wallpaper_set_command: None,
            resize_algorithm: None,
            rwal_params: None,
            image_operations: None,
        }
    }

    pub fn read(&mut self, path: &str) {
        if let Some(value) = read_value(path) {
            self.displays = read_displays(&value);
            self.templates = read_templates(&value);
            self.wallpaper_set_command = read_wallpaper_set_command(&value);
            self.resize_algorithm = read_resize_algorithm(&value);
            self.rwal_params = read_rwal_params(&value);
            self.image_operations = read_image_operations(&value);
        }
    }

    pub fn read_from_string(&mut self, string: String) {
        if let Some(value) = read_value_from_string(string) {
            self.displays = read_displays(&value);
            self.templates = read_templates(&value);
            self.wallpaper_set_command = read_wallpaper_set_command(&value);
            self.resize_algorithm = read_resize_algorithm(&value);
            self.rwal_params = read_rwal_params(&value);
            self.image_operations = read_image_operations(&value);
        }
    }
}

fn read_value(path: &str) -> Option<Value> {
    if let Ok(mut file) = File::open(path) {
        let mut json_data = String::new();
        
        if let Err(_) = file.read_to_string(&mut json_data) {
            return  None;
        }

        if let Ok(data) = serde_json::from_str(&json_data) {
            return data;
        }
    }
    None
}

fn read_value_from_string(string: String) -> Option<Value> {
    if let Ok(data) = serde_json::from_str(&string) {
        return data;
    }
    None
}

fn read_displays(value: &Value) -> Option<Vec<Display>> {
    let mut displays = Vec::new();

    for raw_display in value["displays"].as_array()? {
        let name = raw_display["name"].as_str()?.to_string();
        let w = raw_display["w"].as_u64()? as u32;
        let h = raw_display["h"].as_u64()? as u32;
        let x = raw_display["x"].as_u64()? as u32;
        let y = raw_display["y"].as_u64()? as u32;

        displays.push(Display::new(w, h, x, y, name))
    }

    Some(displays)
}

fn read_templates(value: &Value) -> Option<Vec<Template>> {
    let mut templates = Vec::new();

    for template_path in value["templates"].as_array()? {
        if let Some(path) = template_path.as_str() {
            templates.push(Template::new(&expand_user(path)));
        }
    }

    Some(templates)
}

fn read_wallpaper_set_command(value: &Value) -> Option<String> {
    if let Some(command) = value["wall_command"].as_str() {
        return Some(String::from(command));
    }

    None
}

fn read_resize_algorithm(value: &Value) -> Option<String> {
    if let Some(command) = value["resize_algorithm"].as_str() {
        return Some(String::from(command));
    }

    None
}

fn read_rwal_params(value: &Value) -> Option<RwalParams> {
    let rwal = value.get("rwal")?;

    let thumb_w = rwal["thumb_w"].as_u64().unwrap_or(200) as u32;
    let thumb_h = rwal["thumb_h"].as_u64().unwrap_or(200) as u32;
    let thumb_range = (thumb_w, thumb_h);

    let clamp_min = rwal["clamp_min"].as_f64().unwrap_or(140.0) as f32;
    let clamp_max = rwal["clamp_max"].as_f64().unwrap_or(170.0) as f32;
    let clamp_range = (clamp_min, clamp_max);
    
    let accent_color = rwal["accent_color"].as_u64().unwrap_or(4) as u32;
    let colors = rwal["rwal_colors"].as_u64().unwrap_or(7) as u32;

    let order = match rwal["order_by"].as_str().unwrap_or("h") {
        "s" | "S" => {OrderBy::Saturation},
        "v" | "V" | "b" | "B" => {OrderBy::Brightness},
        _ => {OrderBy::Hue},
    };

    Some(RwalParams::new(thumb_range, clamp_range, accent_color, colors, order))
}

fn read_image_operations(value: &Value) -> Option<ImageOperations> {
    let impg = value.get("impg")?;

    let contrast = impg["contrast"].as_f64().unwrap_or(0.0) as f32;
    let brightness = impg["brightness"].as_i64().unwrap_or(0) as i32;
    let hue = impg["huerotate"].as_i64().unwrap_or(0) as i32;
    let blur = impg["blur"].as_f64().unwrap_or(0.0) as f32;
    let invert = impg["invert"].as_bool().unwrap_or(false);
    let flip_h = impg["flip_h"].as_bool().unwrap_or(false);
    let flip_v = impg["flip_v"].as_bool().unwrap_or(false);

    Some(ImageOperations::new(contrast, brightness, hue, blur, invert, flip_h, flip_v))
}

pub trait JsonString {
    fn json(& self) -> String {
        String::new()
    }
}

impl JsonString for Display {
    fn json(& self) -> String {
        format!(
            "{{\"name\":\"{}\",\"w\":{},\"h\":{},\"x\":{},\"y\":{}}}",
            self.name(),
            self.width(),
            self.height(),
            self.x(),
            self.y(),
        )
    }
}

impl JsonString for Vec<Display> {
    fn json(& self) -> String {
        let json_strings: Vec<String> = self.iter()
            .map(|d| d.json())
            .collect();

        format!("[{}]", json_strings.join(","))
    }
}

impl JsonString for Template {
    fn json(& self) -> String {
        format!(
            "\"{}\"",
            self.path
        )
    }
}

impl JsonString for Vec<Template> {
    fn json(& self) -> String {
        let json_strings: Vec<String> = self.into_iter()
            .map(|t| t.json())
            .collect();

        format!("[{}]", json_strings.join(","))
    }
}

impl JsonString for ImageOperations {
    fn json(& self) -> String {
        format!(
            "{{\"contrast\":{},\"brightness\":{},\"huerotate\":{},\"blur\":{},\"invert\":{},\"flip_h\":{},\"flip_v\":{}}}",
            self.contrast,
            self.brightness,
            self.hue,
            self.blur,
            self.invert,
            self.flip_h,
            self.flip_v
        )
    }
}

impl JsonString for RwalParams {
    fn json(& self) -> String {
        format!(
            "{{\"thumb_w\":{},\"thumb_h\":{},\"accent_color\":{},\"clamp_min\":{},\"clamp_max\":{}}}",
            self.thumb_range.0,
            self.thumb_range.1,
            self.accent_color,
            self.clamp_range.0,
            self.clamp_range.1
        )
    }
}

impl JsonString for Config {
    fn json(& self) -> String {
        format!(
            "{{\"displays\":{},\"templates\":{},\"impg\":{},\"rwal\":{}}}",
            {
                if let Some(displays) = &self.displays {
                    displays.json()
                } else {
                    "[]".to_string()
                }
            },
            {
                if let Some(templates) = &self.templates {
                    templates.json()
                } else {
                    "[]".to_string()
                }
            },
            {
                if let Some(image_operations) = &self.image_operations {
                    image_operations.json()
                } else {
                    "{}".to_string()
                }
            },
            {
                if let Some(rwal_params) = &self.rwal_params {
                    rwal_params.json()
                } else {
                    "{}".to_string()
                }
            }
        )
    }
}

impl JsonString for WCacheInfo {
    fn json(& self) -> String {
        format!(
            "{{\"display\":\"{}\",\"path\":\"{}\"}}",
            self.display_name,
            self.path
        )
    }
}